src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs

```diff
@@ -71,19 +89,194 @@ protected override void EndProcessing()
+                        indexedMinMaxHeap[childIndex] = temp;
+
+                        childIndex = parentIndex;
+                    }
+
+                    // If the heap size is too large, remove the root and rearrange the heap
+                    if (indexedMinMaxHeap.Count > indexedMinMaxHeapSize)
+                    {
+                        // Move the last item added to the root 
+                        indexedMinMaxHeap[0] = indexedMinMaxHeap[indexedMinMaxHeapSize];
+                        indexedMinMaxHeap.RemoveAt(indexedMinMaxHeapSize);
+
+                        // Bubble the root item down into the correct position
+                        int parentIndex = 0;
+                        int parentItemCount = indexedMinMaxHeap.Count >> 1;
+                        while (parentIndex < parentItemCount)
+                        {
+                            // Min-heap: use the smaller of the two children in the comparison
+                            // Max-heap: use the larger of the two children in the comparison
+                            int leftChildIndex = (parentIndex << 1) + 1;
+                            int rightChildIndex = leftChildIndex + 1;
+                            childIndex = rightChildIndex == indexedMinMaxHeap.Count || comparer.Compare(indexedMinMaxHeap[leftChildIndex], indexedMinMaxHeap[rightChildIndex]) != comparator
+                                ? leftChildIndex
+                                : rightChildIndex;
+
+                            // Min-heap: if the smallest child is larger than or equal to its parent, break
+                            // Max-heap: if the largest child is smaller than or equal to its parent, break
+                            int childComparisonResult = comparer.Compare(indexedMinMaxHeap[childIndex], indexedMinMaxHeap[parentIndex]); 
+                            if (childComparisonResult == 0 || childComparisonResult == comparator)
+                            {
+                                break;
+                            }
+
+                            var temp = indexedMinMaxHeap[childIndex];
+                            indexedMinMaxHeap[childIndex] = indexedMinMaxHeap[parentIndex];
+                            indexedMinMaxHeap[parentIndex] = temp;
+
+                            parentIndex = childIndex;
+                        }
+                    }
+                }
+
+                indexedMinMaxHeap.Sort(comparer);
+
+                foreach (var entry in indexedMinMaxHeap)
+                {
+                    WriteObject(entry.Entry.inputObject);
+                }
             }
+        }
+    }
 
-            // write to output stream
-            foreach (OrderByPropertyEntry x in orderByProperty.OrderMatrix)
+    /// <summary>
+    /// This is a single entry in the min-/max-heap used for top N/bottom N sorts
+    /// </summary>
+    public sealed class IndexedMinMaxHeapEntry
```

Comment:
Changed locally, will commit with other changes soon. On the `struct` comment, should this be a struct as well then, for the performance win?
```C#
    internal sealed class OrderByPropertyEntry
    {
        internal PSObject inputObject = null;
        internal List<ObjectCommandPropertyValue> orderValues = new List<ObjectCommandPropertyValue>();
    }
```